src/codecs/tiff.rs

```diff
@@ -127,20 +130,47 @@ where
     }
 }
 
-fn check_sample_format(sample_format: u16) -> Result<(), ImageError> {
-    match tiff::tags::SampleFormat::from_u16(sample_format) {
-        Some(tiff::tags::SampleFormat::Uint) => Ok(()),
-        Some(other) => Err(ImageError::Unsupported(
+fn check_sample_format(sample_format: u16, color_type: tiff::ColorType) -> Result<(), ImageError> {
+    use tiff::{tags::SampleFormat, ColorType};
+    let num_bits = match color_type {
+        ColorType::CMYK(k) => k,
+        ColorType::Gray(k) => k,
+        ColorType::RGB(k) => k,
+        ColorType::RGBA(k) => k,
+        ColorType::GrayA(k) => k,
+        ColorType::Palette(k) | ColorType::YCbCr(k) => {
+            return Err(ImageError::Unsupported(
+                UnsupportedError::from_format_and_kind(
+                    ImageFormat::Tiff.into(),
+                    UnsupportedErrorKind::GenericFeature(format!(
+                        "Unhandled TIFF color type {color_type:?} for {k} bits",
+                    )),
+                ),
+            ))
+        }
+        _ => {
+            return Err(ImageError::Unsupported(
+                UnsupportedError::from_format_and_kind(
+                    ImageFormat::Tiff.into(),
+                    UnsupportedErrorKind::GenericFeature(format!(
+                        "Unhandled TIFF color type {color_type:?}",
+                    )),
+                ),
+            ))
+        }
+    };
+
+    match SampleFormat::from_u16(sample_format) {
+        Some(SampleFormat::Uint) if num_bits <= 16 => Ok(()),
+        Some(SampleFormat::IEEEFP) if num_bits == 32 => Ok(()),
```

Comment:
I don't love the implicitness of "num_bits <= 16 means integer" and "num_bits == 32 means float". But really this should be addressed by having the `tiff` crate expose a way to directly get the sample format, so I'm OK with doing it this way for now.

Reply:
Implicit is probably the wrong word. The concern is that the `tiff::ColorType::RGB(32) => ColorType::Rgb32F` line above only works because of the checks that are happening here. The result is you have to read the implementation of `check_sample_format` to understand what is being checked, and only afterwards can you understand why `new` can match on color type without checking the sample format.

If `new` had access to both the sample format and sample type, the logic could be more straightforward:

```rust
match sample_format {
    SampleFormat::Uint => match color_type {
        RGB(8) => ColorType::Rgb8,
        ...
    },
    SampleFormat::IEEEFP => match color_type {
        RGB(32) => ColorType::Rgb32F
        ...
    }
}
```